picard/util/imagelist.py

```diff
@@ -35,73 +36,156 @@ def __getitem__(self, k):
+        obj.orig_metadata.has_common_images = state.has_common_orig_images
 
 
 # TODO: use functools.singledispatch when py3 is supported
-def update_metadata_images(obj):
-    from picard.track import Track
-    from picard.cluster import Cluster
+def _get_state(obj):
     from picard.album import Album
+    from picard.cluster import Cluster
+    from picard.track import Track
 
-    state = State()
+    state = ImageListState()
 
     if isinstance(obj, Album):
-        sources = []
         for track in obj.tracks:
-            sources.append(track)
-            sources += track.linked_files
-        sources += obj.unmatched_files.files
+            state.sources.append(track)
+            state.sources += track.linked_files
+        state.sources += obj.unmatched_files.files
         state.update_new_metadata = True
         state.update_orig_metadata = True
     elif isinstance(obj, Track):
-        sources = obj.linked_files
+        state.sources = obj.linked_files
         state.update_orig_metadata = True
     elif isinstance(obj, Cluster):
-        sources = obj.files
+        state.sources = obj.files
         state.update_new_metadata = True
 
-    for src_obj in sources:
-        _process_images(state, src_obj)
+    return state
+
+
+def _get_metadata_images(state, sources):
+    new_images = set()
+    orig_images = set()
+    for s in sources:
+        if state.update_new_metadata:
+            new_images = new_images.union(s.metadata.images)
+        if state.update_orig_metadata:
+            orig_images = orig_images.union(s.orig_metadata.images)
+    return (new_images, orig_images)
+
+
+def update_metadata_images(obj):
+    _update_state(obj, _get_state(obj))
+
+
+def _add_images(metadata, added_images):
+    if not added_images:
+        return
+
+    current_images = set(metadata.images)
+    if added_images == current_images:
+        return
+    else:
+        metadata.images = ImageList(current_images.union(added_images))
+        metadata.has_common_images = False
+
+
+def add_metadata_images(obj, added_sources):
+    state = _get_state(obj)
+    (added_new_images, added_orig_images) = _get_metadata_images(state, added_sources)
 
     if state.update_new_metadata:
-        obj.metadata.images = state.new_images
-        obj.metadata.has_common_images = state.has_common_new_images
+        _add_images(obj.metadata, added_new_images)
+    if state.update_orig_metadata:
+        _add_images(obj.orig_metadata, added_orig_images)
+
+
+def _remove_images(metadata, sources, removed_images):
+    if not metadata.images or not removed_images:
+        return
+
+    if not sources:
+        metadata.images = ImageList()
+        metadata.has_common_images = True
+        return
+
+    current_images = set(metadata.images)
+
+    if metadata.has_common_images and current_images == removed_images:
+        return
+
+    common_images = True
+    previous_images = None
+    for source_metadata in sources:
```

Comment:
I think a comment explaining a bit what's the actual process could help, the code is clean but not really easy to grasp imho.
Perhaps just defining what are `source_images`, `previous_images`, `common_images`, `removed_images`, dunno